File: design_patterns/08_Observer/broadcast.py
from abc import ABC, abstractmethod

# Abstract Subject
class Subject(ABC):
    @abstractmethod
    def register_observer(self, observer):
        pass
    
    @abstractmethod
    def remove_observer(self, observer):
        pass
    
    @abstractmethod
    def notify_observers(self):
        pass

# Concrete Subject
class WeatherStation(Subject):
    def __init__(self):
        self._temperature = None
        self._observers = []

    def register_observer(self, observer):
        self._observers.append(observer)

    def remove_observer(self, observer):
        self._observers.remove(observer)

    def notify_observers(self):
        for observer in self._observers:
            observer.update(self._temperature)

    def set_temperature(self, temperature):
        print("Setting temperature to:", temperature)
        self._temperature = temperature
        self.notify_observers()

# Abstract Observer
class Observer(ABC):
    @abstractmethod
    def update(self, temperature):
        pass

# Concrete Observer
class Display(Observer):
    def __init__(self, name):
        self._name = name

    def update(self, temperature):
        print(f"{self._name} received temperature update: {temperature}")

# Usage example
if __name__ == "__main__":
    weather_station = WeatherStation()

    observer1 = Display("Observer 1")
    observer2 = Display("Observer 2")

    weather_station.register_observer(observer1)
    weather_station.register_observer(observer2)

    weather_station.set_temperature(25.5)
    weather_station.set_temperature(30.0)

    weather_station.remove_observer(observer1)

    weather_station.set_temperature(27.8)

File: design_patterns/08_Observer/broadcast.js
// Subject (Observable)

// Libraries = Redux? Sveltestore? 
class WeatherStation {
    constructor() {
        this.temperature = null; // Variables (states) are dynamically taken care of, each state => observers
        this.observers = [];
    }

    registerObserver(observer) {
        this.observers.push(observer);
    }

    removeObserver(observer) {
        const index = this.observers.indexOf(observer);
        if (index !== -1) {
            this.observers.splice(index, 1);
        }
    }

    notifyObservers() {
        this.observers.forEach(observer => observer.update(this.temperature));
    }

    setTemperature(temperature) {
        console.log(`Setting temperature to: ${temperature}`);
        this.temperature = temperature;
        this.notifyObservers();
    }
}

// Observer
class Display {
    constructor(name) {
        this.name = name;
        // temperature
    }

    update(temperature) {
        console.log(`${this.name} received temperature update: ${temperature}`);
        // this.temperature = temperature
        // Whatever you want to trigger, when value change happens.
    }
}

// Usage example
const weatherStation = new WeatherStation();

const observer1 = new Display("Observer 1");
const observer2 = new Display("Observer 2");

weatherStation.registerObserver(observer1);
weatherStation.registerObserver(observer2);

weatherStation.setTemperature(25.5);
weatherStation.setTemperature(30.0);

weatherStation.removeObserver(observer1);

weatherStation.setTemperature(27.8);
